### PR TITLE
Keras callback to push to hub each epoch, or after N steps

### PR DESCRIPTION
**Pull request by Rocketknight1**

### CODE REVIEW

File: src/transformers/keras_callbacks.py [hunk 1]
```diff
@@ -0,0 +1,36 @@
+from typing import Optional
+
+from tensorflow.keras.callbacks import Callback
+
+
+class PushToHubCallback(Callback):
+    def __init__(
+        self,
+        save_strategy="epoch",
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:9]: You can default to "epoch" but you should use the `IntervalStrategy` for this arg (and cast it from the string to that type). Also we need a type annotation
```suggestion
        save_strategy: Union[str, IntervalStrategy] = "epoch",
```

File: src/transformers/keras_callbacks.py [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+        save_steps: Optional[int] = None,
+        repo_path_or_name: Optional[str] = None,
+        repo_url: Optional[str] = None,
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:12]: I would prefer we use the same arguments as the `Trainer`: `output_dir`, `hub_model_id` and `hub_token` to be consistent between TF and PyTorch.

File: src/transformers/keras_callbacks.py [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+        **push_to_hub_kwargs
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:13]: Not sure we need to show those: the blocking argument should be set to False so training continues, and the commit messages should be different per step.

Also the tokenizer should be accepted as an optional argument.

File: src/transformers/keras_callbacks.py [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+        if repo_path_or_name is None and repo_url is None:
+            raise ValueError("You need to specify a `repo_path_or_name` or a `repo_url`.")
+        self.repo_path_or_name = repo_path_or_name
+        self.repo_url = repo_url
+        self.push_to_hub_kwargs = push_to_hub_kwargs
+
+    def on_train_batch_end(self, batch, logs=None):
+        if self.save_strategy == "steps" and batch + 1 % self.save_steps == 0:
+            self.model.push_to_hub(
+                repo_path_or_name=self.repo_path_or_name, repo_url=self.repo_url, **self.push_to_hub_kwargs
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:29]: The commit message should be "Training in progress, step xxx`

File: src/transformers/keras_callbacks.py [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+        self.save_strategy = save_strategy
+        if self.save_strategy == "steps" and (not isinstance(save_steps, int) or save_steps <= 0):
+            raise ValueError("Please supply a positive integer argument for save_steps when save_strategy == 'steps'!")
+        self.save_steps = save_steps
+        if repo_path_or_name is None and repo_url is None:
+            raise ValueError("You need to specify a `repo_path_or_name` or a `repo_url`.")
+        self.repo_path_or_name = repo_path_or_name
+        self.repo_url = repo_url
+        self.push_to_hub_kwargs = push_to_hub_kwargs
+
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:25]: We should set a repository object at init.

File: src/transformers/keras_callbacks.py [hunk 6]
```diff
@@ -0,0 +1,36 @@
...
+    def on_train_batch_end(self, batch, logs=None):
+        if self.save_strategy == "steps" and batch + 1 % self.save_steps == 0:
+            self.model.push_to_hub(
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:28]: I would prefer we use `self.model.save_pretrained`, save the optional tokenizer and then use `repo.push_to_hub`

File: src/transformers/keras_callbacks.py [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+                repo_path_or_name=self.repo_path_or_name, repo_url=self.repo_url, **self.push_to_hub_kwargs
+            )
+
+    def on_epoch_end(self, epoch, logs=None):
+        if self.save_strategy == "epoch":
+            self.model.push_to_hub(
+                repo_path_or_name=self.repo_path_or_name, repo_url=self.repo_url, **self.push_to_hub_kwargs
+            )
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:36]: Similar comments here.

Also add a `on_train_end` last commit?

**Review by sgugger — COMMENTED** [review]
I have many comments :-)
This is taking the old approach we used in Trainer and with which we were not satisfied as it created lots of problems downstream.

We should really follow the road using a `Repository` object as it's closer to a git workflow and is less surprising for us or the user.

File: src/transformers/keras_callbacks.py (hunk 1 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Done!

File: src/transformers/keras_callbacks.py (hunk 3 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Done!

File: src/transformers/keras_callbacks.py [hunk 8]
```diff
@@ -0,0 +1,65 @@
+from pathlib import Path
+from typing import Optional, Union
+
+from tensorflow.keras.callbacks import Callback
+
+from huggingface_hub import Repository
+from transformers.file_utils import get_full_repo_name
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:7]:
```suggestion
from .file_utils import get_full_repo_name
```
Let's use relative imports :-)

File: src/transformers/keras_callbacks.py [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+    ):
+        super().__init__()
+        if isinstance(save_strategy, str):
+            save_strategy = IntervalStrategy(save_strategy.lower())
+        self.save_strategy = save_strategy
+        if self.save_strategy == IntervalStrategy.STEPS and (not isinstance(save_steps, int) or save_steps <= 0):
+            raise ValueError("Please supply a positive integer argument for save_steps when save_strategy == 'steps'!")
+        self.save_steps = save_steps
+        if isinstance(output_dir, str):
+            output_dir = Path(output_dir)
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:30]:
```suggestion
        output_dir = Path(output_dir)
```
Nit, `Path` does that for you already.

File: src/transformers/keras_callbacks.py [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+        if hub_model_id is None:
+            repo_name = get_full_repo_name(output_dir.name, token=hub_token)
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:32]:
```suggestion
            repo_name = get_full_repo_name(output_dir.absolute().name, token=hub_token)
```
You need to resolve the path in case the user passes something like `"."`

File: src/transformers/keras_callbacks.py [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+        self.repo = Repository(str(output_dir), clone_from=repo_name)
+        self.tokenizer = tokenizer
+        self.last_job = None
+
+    def on_train_batch_end(self, batch, logs=None):
+        if self.save_strategy == IntervalStrategy.STEPS and batch + 1 % self.save_steps == 0:
+            if self.last_job is not None and not self.last_job.is_done():
+                return  # The last upload is still running, don't start another
+            self.model.save_pretrained(self.output_dir)
+            self.tokenizer.save_pretrained(self.output_dir)
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:45]:
```suggestion
            if self.tokenizer is not None:
                self.tokenizer.save_pretrained(self.output_dir)
```

File: src/transformers/keras_callbacks.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+            _, self.last_job = self.repo.push_to_hub(
+                commit_message=f"Training in progress steps {batch}", blocking=False
+            )
+
+    def on_epoch_end(self, epoch, logs=None):
+        if self.save_strategy == IntervalStrategy.EPOCH:
+            if self.last_job is not None and not self.last_job.is_done():
+                return  # The last upload is still running, don't start another
+            self.model.save_pretrained(self.output_dir)
+            self.tokenizer.save_pretrained(self.output_dir)
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:55]:
```suggestion
            if self.tokenizer is not None:
                self.tokenizer.save_pretrained(self.output_dir)
```

File: src/transformers/keras_callbacks.py [hunk 13]
```diff
@@ -0,0 +1,65 @@
...
+            _, self.last_job = self.repo.push_to_hub(
+                commit_message=f"Training in progress epoch {epoch}", blocking=False
+            )
+
+    def on_train_end(self, logs=None):
+        if self.last_job is not None and not self.last_job.is_done():
+            self.last_job.process.join()  # Wait for existing upload to finish to finish
+        self.model.save_pretrained(self.output_dir)
+        self.tokenizer.save_pretrained(self.output_dir)
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:64]:
```suggestion
            if self.tokenizer is not None:
                self.tokenizer.save_pretrained(self.output_dir)
```

File: src/transformers/keras_callbacks.py [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+                return  # The last upload is still running, don't start another
+            self.model.save_pretrained(self.output_dir)
+            self.tokenizer.save_pretrained(self.output_dir)
+            _, self.last_job = self.repo.push_to_hub(
+                commit_message=f"Training in progress epoch {epoch}", blocking=False
+            )
+
+    def on_train_end(self, logs=None):
+        if self.last_job is not None and not self.last_job.is_done():
+            self.last_job.process.join()  # Wait for existing upload to finish to finish
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:62]: Maybe print something so the user is aware we are waiting for it?

**Review by sgugger — APPROVED** [review]
Thanks a lot for updating!

File: src/transformers/keras_callbacks.py (hunk 14 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Done!

File: src/transformers/keras_callbacks.py [hunk 15]
```diff
@@ -0,0 +1,70 @@
...
+            self.model.save_pretrained(self.output_dir)
+            if self.tokenizer is not None:
+                self.tokenizer.save_pretrained(self.output_dir)
+            _, self.last_job = self.repo.push_to_hub(
+                commit_message=f"Training in progress epoch {epoch}", blocking=False
+            )
+
+    def on_train_end(self, logs=None):
+        if self.last_job is not None and not self.last_job.is_done:
+            print("Waiting for existing upload to finish...")
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:64]: Use the internal logger instead of a print?

**Review by Hecim1984 — APPROVED** [review]

**Comment by Hecim1984** [comment]
🚀🎉

**Review by Hecim1984 — CHANGES_REQUESTED** [review]
Help

File: src/transformers/keras_callbacks.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        hub_model_id: Optional[str] = None,
+        hub_token: Optional[str] = None,
+    ):
+        """
+        output_dir (:obj:`str`):
+            The output directory where the model predictions and checkpoints will be written. The name of the
+            repository to keep in sync with the local `output_dir`. Should be the whole repository
+        save_strategy (:obj:`str` or :class:`~transformers.trainer_utils.IntervalStrategy`,
+        `optional`, defaults to :obj:`"epoch"`):
```
Comment (sgugger) [inline src/transformers/keras_callbacks.py:28]: This need to be on the same line cause sphinx sucks.

**Review by sgugger — APPROVED** [review]

File: src/transformers/keras_callbacks.py [hunk 17]
```diff
@@ -0,0 +1,94 @@
...
+        save_strategy: Union[str, IntervalStrategy] = "epoch",
+        save_steps: Optional[int] = None,
+        tokenizer: Optional[PreTrainedTokenizerBase] = None,
+        hub_model_id: Optional[str] = None,
+        hub_token: Optional[str] = None,
+    ):
+        """
+        output_dir (:obj:`str`):
+            The output directory where the model predictions and checkpoints will be written. The name of the
+            repository to keep in sync with the local `output_dir`. Should be the whole repository
```
Comment (LysandreJik) [inline src/transformers/keras_callbacks.py:26]: I had a hard time understanding that docstring as it gives what seems to be unrelated information.

File: src/transformers/keras_callbacks.py [hunk 18]
```diff
@@ -0,0 +1,94 @@
...
+        save_strategy (:obj:`str` or :class:`~transformers.trainer_utils.IntervalStrategy`,
+        `optional`, defaults to :obj:`"epoch"`):
+            The checkpoint save strategy to adopt during training. Possible values are:
+
+                * :obj:`"no"`: No save is done during training.
+                * :obj:`"epoch"`: Save is done at the end of each epoch.
+                * :obj:`"steps"`: Save is done every :obj:`save_steps`
+        save_steps (:obj:`int`, `optional`):
+            The number of steps between saves when using the "steps" save_strategy.
+        tokenizer: (:obj:`PreTrainedTokenizerBase`, `optional`):
```
Comment (LysandreJik) [inline src/transformers/keras_callbacks.py:36]:
```suggestion
        tokenizer (:obj:`PreTrainedTokenizerBase`, `optional`):
```

**Review by LysandreJik — APPROVED** [review]
This looks good, thank you for working on it @Rocketknight1!

File: src/transformers/keras_callbacks.py (hunk 18 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py:36]: Done!

File: src/transformers/keras_callbacks.py (hunk 16 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Done!

File: src/transformers/keras_callbacks.py (hunk 17 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Agreed, I copy-pasted it from elsewhere but it feels a bit incomplete - I might have missed something or mangled it somehow. I rewrote it.

File: src/transformers/keras_callbacks.py (hunk 15 above)
Comment (Rocketknight1) [reply src/transformers/keras_callbacks.py]: Done!
